### PR TITLE
Feature/refine flag preprocessing

### PR DESCRIPTION
**Pull request by drmacro**
Creating pull request to serve as issue.

Basic problem is that the filterImpl.xsl is implemented using nasty XSLT 1 approaches that do not correctly handle grouped values. Have started refactoring to use XSLT 2 features and correctly implement the grouping spec.

I'm using this as my primary test case: 
https://github.com/dita-community/dita-test-cases/blob/master/flagging-test/conditions_with_groups.dita

Will indicate when I feel the functionality is ready to merge. This is targeted for 2.3

### CODE REVIEW

File: src/main/xsl/preprocess/flagImpl.xsl [hunk 1]
```diff
@@ -65,20 +65,26 @@ LOOK FOR FIXME TO FIX SCHEMEDEF STUFF
...
+    
+    <xsl:variable name="propsAttrs" as="xs:string*">
+      <xsl:analyze-string select="$GLOBAL-DOMAINS" regex="a\(props\s+(\w+)\.*?\)">
+        <xsl:matching-substring>
+          <xsl:sequence select="regex-group(1)"/>
+        </xsl:matching-substring>
+      </xsl:analyze-string>
     </xsl:variable>
-    <xsl:value-of select="substring-after($collectPropsExtensions, ',')"/>
+<!--    <xsl:message> + [DEBUG] propsExtensions="<xsl:value-of select="$propsAttrs"/>"</xsl:message>-->
```
Comment (jelovirt) [inline src/main/xsl/preprocess/flagImpl.xsl]: Remove

File: src/main/xsl/preprocess/flagImpl.xsl [hunk 2]
```diff
@@ -337,277 +334,146 @@ LOOK FOR FIXME TO FIX SCHEMEDEF STUFF
...
+         be specified multiple times on a single element and you don't want multiple flag indicators in that case.
+         -->
+        <xsl:variable name="conditions" as="element()">
+          <conditions>
+            <xsl:apply-templates mode="getconditions"
+              select="@audience, @platform, @product, @otherprops, @props, @rev, @*[name(.) = $propsExtensions]"
+            />
+          </conditions>
+        </xsl:variable>        
+        <xsl:message> + [DEBUG] getrules: conditions:
```
Comment (jelovirt) [inline src/main/xsl/preprocess/flagImpl.xsl]: Remove

File: src/main/xsl/preprocess/flagImpl.xsl [hunk 3]
```diff
@@ -337,277 +334,146 @@ LOOK FOR FIXME TO FIX SCHEMEDEF STUFF
...
-      <xsl:otherwise>
-        <xsl:if test="contains($flag-att-path,' ')">
-          <xsl:call-template name="ext-gen-prop">
-            <xsl:with-param name="flag-att-path" select="normalize-space(substring-before($flag-att-path, $propName))"/>
-            <xsl:with-param name="flag-att-val" select="$flag-att-val"/>
-          </xsl:call-template>
-        </xsl:if>
-      </xsl:otherwise>
-    </xsl:choose>
+<!--    <xsl:message> + [DEBUG] getconditions: @<xsl:value-of select="name(.)"/>: "<xsl:value-of select="."/>"</xsl:message>-->
```
Comment (jelovirt) [inline src/main/xsl/preprocess/flagImpl.xsl]: Remove

File: src/main/xsl/preprocess/flagImpl.xsl [hunk 4]
```diff
@@ -337,277 +334,146 @@ LOOK FOR FIXME TO FIX SCHEMEDEF STUFF
...
+    <xsl:if test="count($bareValues) gt 0">
+      <condition name="{name(.)}">
+        <xsl:for-each select="distinct-values($bareValues)">
+          <value><xsl:sequence select="."/></value>
+        </xsl:for-each>
+      </condition>
+    </xsl:if>
+    
+    <!-- Making this check so we don't accidently support
```
Comment (jelovirt) [inline src/main/xsl/preprocess/flagImpl.xsl]: Remove

File: src/main/xsl/preprocess/flagImpl.xsl [hunk 5]
```diff
@@ -337,277 +334,146 @@ LOOK FOR FIXME TO FIX SCHEMEDEF STUFF
...
-    <xsl:copy-of select="$FILTERDOC/val/revprop[@val=$firstflag][1][@action='flag']"/>
+    <xsl:sequence select="$FILTERDOC/val/revprop[@val=$firstflag][1][@action='flag']"/>
    </xsl:when>
    <xsl:when test="$FILTERDOC/val/prop[@att=$flag-att][@val=$firstflag][1][@action='flag']">
-    <xsl:copy-of select="$FILTERDOC/val/prop[@att=$flag-att][@val=$firstflag][1][@action='flag']"/>
+    <xsl:sequence select="$FILTERDOC/val/prop[@att=$flag-att][@val=$firstflag][1][@action='flag']"/>
    </xsl:when>
    <xsl:when test="$FILTERDOC/val/prop[@att=$flag-att][not(@val=$firstflag)][@action='flag']">
     
+    <!-- FIXME: WEK: This should be done with apply-templates. -->
```
Comment (jelovirt) [inline src/main/xsl/preprocess/flagImpl.xsl]: Remove

File: src/main/xsl/preprocess/flagImpl.xsl [hunk 6]
```diff
@@ -65,20 +65,26 @@ LOOK FOR FIXME TO FIX SCHEMEDEF STUFF
...
-    <xsl:value-of select="substring-after($collectPropsExtensions, ',')"/>
+<!--    <xsl:message> + [DEBUG] propsExtensions="<xsl:value-of select="$propsAttrs"/>"</xsl:message>-->
+    <xsl:sequence select="$propsAttrs"/>
   </xsl:variable>
+  
 
   <xsl:template match="/">
     <!-- Avoid all later checks by adding test here - if no filter file, copy full tree? -->
-    <xsl:apply-templates/>
+    <xsl:apply-templates/>
```
Comment (jelovirt) [inline src/main/xsl/preprocess/flagImpl.xsl]: Whitespace bug

File: src/main/xsl/preprocess/flagImpl.xsl [hunk 7]
```diff
@@ -87,51 +93,50 @@ LOOK FOR FIXME TO FIX SCHEMEDEF STUFF
...
-              <xsl:with-param name="flagrules" select="$flagrules"/>
+              <xsl:with-param name="flagrules" as="element()*" select="$flagrules"/>
             </xsl:apply-templates>
           </ditaval-endprop>
         </xsl:copy>
       </xsl:when>
       <xsl:otherwise>
         <xsl:copy>
-          <xsl:apply-templates select="@*|*|processing-instruction()|comment()|text()"/>
+          <xsl:apply-templates select="@*,node()"/>
```
Comment (jelovirt) [inline src/main/xsl/preprocess/flagImpl.xsl]: Prefer `@* | node()` to `@*, node()`, since that's been the convention in the other stylesheets too.

**Comment by drmacro** [comment]
@*, node() is the correct way to do it in XSLT 2 since it ensures that all attributes will be emitted before any nodes.

That is, it's the standard idiom for identity transforms in XSLT 2. Existing practice is wrong and should be fixed.

**Comment by robander** [comment]
I agree that `@*, node()` is better practice but can't help but thinking ... `@* | node()` has always worked, everybody is familiar with it, and there's not a big incentive to change them globally.

**Comment by jelovirt** [comment]
Per the [XPath datamodel](https://www.w3.org/TR/xpath-datamodel/#document-order) attributes come before child nodes, thus `@* | node()` will always result in the same order as `@*, node()`.

**Comment by drmacro** [comment]
You could have a sequence of mixed attribute nodes and other nodes to which you apply templates. In that case "node() | @*" would fail.

**Comment by jelovirt** [comment]
Union operator `|` will alway return the nodes in document node, and since `@*` and `node()` will always have the same parent, current node, `node() | @*` will always return attributes before child nodes.

**Comment by jelovirt** [comment]
@drmacro could you enable [Allow edits from maintainers](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) on this PR?

**Comment by jelovirt** [comment]
Reimplementing flagging as part of filtering in #2748 has made this PR obsolete.

**Pull request closed without merging**